### PR TITLE
Javascript additional functionality

### PR DESCRIPTION
**Pull request by wolfv**
This adds some functionality to the Javascript port:

- two way JS <-> Python communication using a js module in Python and emscripten eval / evali functionality
- adds code to wrapper.js to initialize the browser correctly, and find all `<script type="script/micropython">` tags and executes them
- another make target `make server` which builds and shows the `test.html` webpage to see a small example of micropython functionality in the browser

### CODE REVIEW

**Comment by pfalcon** [comment]
@wolfv, Did you see https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md ?

**Comment by wolfv** [comment]
I guess you were referring to the commit message style? Changed it.

File: ports/javascript/Makefile [hunk 1]
```diff
@@ -36,6 +36,7 @@ SRC_C = \
 	main.c \
 	mphalport.c \
 	modutime.c \
+	js_module.c \
```
Comment (dpgeorge) [inline ports/javascript/Makefile]: Following how it's done elsewhere, this should be called modjs.c

File: ports/javascript/README.md [hunk 2]
```diff
@@ -7,16 +7,16 @@ Dependencies
 ------------
 
 Building micropython.js bears the same requirements as the standard MicroPython
-ports with the addition of Emscripten (and uglify-js for the minified file). 
+ports with the addition of Emscripten (and uglify-js for the minified file).
```
Comment (dpgeorge) [inline ports/javascript/README.md]: Usually, unrelated changes like fixing things like this whitespace would be done in separate commits, to keep changes self contained and easier to review/understand.

File: ports/javascript/hello.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import js
```
Comment (dpgeorge) [inline ports/javascript/hello.py]: Is this file needed at all?  If so (as a simple example) it should probably go in the examples/ dir, under a new examples/javascript/ subdir.

File: ports/javascript/js_module.c [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#include "py/nlr.h"
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "py/binary.h"
+
+#include <stdio.h>
+
+// forward declare emscripten stuff
+void emscripten_run_script(const char *);
+int emscripten_run_script_int(const char *);
```
Comment (dpgeorge) [inline ports/javascript/js_module.c]: Does emscripten have a header file for these?

File: ports/javascript/js_module.c [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+STATIC mp_obj_t js_eval(mp_obj_t js_string) {
+    emscripten_run_script(mp_obj_str_get_str(js_string));
+    return mp_const_none;
+}
+
+STATIC mp_obj_t js_evali(mp_obj_t js_string) {
```
Comment (dpgeorge) [inline ports/javascript/js_module.c]: Are there any other return types that could be supported, like float?  They don't have to be added now, but good to think how they would be consistently named (eg evalf) if they were added in the future.

File: ports/javascript/wrapper.js [hunk 6]
```diff
@@ -71,7 +71,42 @@ var mainProgram = function()
...
+var __initBrowserCount = 2;
+
+function initBrowser(e)
+{
+  __initBrowserCount--;
+  // window.onLoad and wasm have loaded...
+  if (__initBrowserCount == 0)
+  {
+    var mp_js_stdout = document.getElementById('mp_js_stdout');
+    mp_js_stdout.addEventListener('print', function(e) {
```
Comment (dpgeorge) [inline ports/javascript/wrapper.js]: I think the reason this was done in the html before is so it can be customised for the particular use/app.  I guess this default behaviour can still be overridden by a particular app, by changing the event listener?

File: ports/javascript/wrapper.js [hunk 7]
```diff
@@ -71,7 +71,42 @@ var mainProgram = function()
...
+      if (e.data == '\n') {
+        mp_js_stdout.innerHTML += "<br/>"
+      } else {
+        mp_js_stdout.innerHTML += e.data;
+      }
+    }, false);
+
+    mp_js_init(64 * 1024);
```
Comment (dpgeorge) [inline ports/javascript/wrapper.js]: This was done in the html previously so it could be changed based on need of the app, eg made bigger or smaller.  How could it be made customisable?

File: ports/javascript/wrapper.js [hunk 8]
```diff
@@ -71,7 +71,42 @@ var mainProgram = function()
...
+    for (var i = 0; i < document.scripts.length; i++)
+    {
+      if (document.scripts[i].type == 'script/micropython')
+      {
+        mp_js_do_str(document.scripts[i].innerText);
+      }
+    }
```
Comment (dpgeorge) [inline ports/javascript/wrapper.js]: Please use the same coding/brace style as already exists in this file.

**Review by dpgeorge — COMMENTED** [review]
Thanks for this, it makes it more usable.  Please see comments in the PR.

**Comment by dpgeorge** [comment]
Closing due to inactivity.  Feel free to reopen with review comments addressed.
